public/app/features/explore/state/explorePane.ts

```diff
@@ -159,50 +160,30 @@ export function initializeExplore(
     if (instance) {
       dispatch(runQueries(exploreId));
     }
-  };
-}
-
-/**
- * Save local redux state back to the URL. Should be called when there is some change that should affect the URL.
- * Not all of the redux state is reflected in URL though.
- */
-export const stateSave = (): ThunkResult<void> => {
-  return (dispatch, getState) => {
-    const { left, right, split } = getState().explore;
-    const orgId = getState().user.orgId.toString();
-    const replace = left && left.urlReplaced === false;
-    const urlStates: { [index: string]: string } = { orgId };
-    urlStates.left = serializeStateToUrlParam(getUrlStateFromPaneState(left), true);
-    if (split) {
-      urlStates.right = serializeStateToUrlParam(getUrlStateFromPaneState(right), true);
-    }
 
-    dispatch(updateLocation({ query: urlStates, replace }));
-    if (replace) {
-      dispatch(setUrlReplacedAction({ exploreId: ExploreId.left }));
-    }
+    const richHistory = getRichHistory();
+    dispatch(richHistoryUpdatedAction({ richHistory }));
   };
-};
+}
 
 /**
- * Reacts to changes in URL state that we need to sync back to our redux state. Checks the internal update variable
- * to see which parts change and need to be synced.
- * @param exploreId
+ * Reacts to changes in URL state that we need to sync back to our redux state. Computes diff of newUrlQuery vs current
+ * state and runs update actions for relevant parts.
  */
-export function refreshExplore(exploreId: ExploreId): ThunkResult<void> {
-  return (dispatch, getState) => {
-    const itemState = getState().explore[exploreId];
+export function refreshExplore(exploreId: ExploreId, newUrlQuery: string): ThunkResult<void> {
+  return async (dispatch, getState) => {
+    const itemState = getState().explore[exploreId]!;
     if (!itemState.initialized) {
       return;
     }
 
-    const { urlState, update, containerWidth, eventBridge } = itemState;
+    // Get diff of what should be updated
+    const newUrlState = parseUrlState(newUrlQuery);
+    const update = urlDiff(newUrlState, getUrlStateFromPaneState(itemState));
```

Comment:
This computes the same state as we had before inside the store state but is here adhoc after we know the url changed. We can still actually not do anything here but that usually should not happen (maybe query reordering or something like that)